Introducing Application Telemetry Schema in OpenTelemetry - Vision and Roadmap (#243)
open-telemetry/oteps: text/0240-app-telemetry-schema-vision-roadmap.md

existing Telemetry Schema v1.1. It offers a way to import semantic convention
registries and to define resource attributes, the properties of the
instrumentation library, and the telemetry signals an application or library can
produce. The existing versioning mechanism defined in version 1.1 remains
unchanged. A Telemetry Schema can inherit from one or several other telemetry
schemas (as explained in the previous section). The root schema is usually the
official Telemetry Schema, referencing the official OpenTelemetry Semantic
Convention Registry. The last schema (leaf) in this hierarchy is called the
Application Telemetry Schema, as it defines all the signals a particular
application produces. Although there is no direct lineage between these systems,
a similar approach was designed and deployed by Facebook to address the same
type of problem but in a proprietary context (refer to this
[positional paper](https://research.facebook.com/publications/positional-paper-schema-first-application-telemetry/)
for more information).

The following diagram illustrates the relationships between these concepts and
the main changes from Telemetry Schema v1.1 to v1.2.

> ![Telemetry Schema](./img/0240-telemetry-schema-v1.1-to-v1.2.png)

Reviewer: A question regarding this diagram: it shows resources and signals are defined as "siblings" of versions. So if I understand correctly, the schema defines resources and signals for the latest version in the schema, and then additional versioned transformations to transform attributes of older versions.

I wonder if there's value in defining those resources and signals for every version, as also those definitions can change across versions? Or asked in a different way: do we want to be able to create SDKs for each version contained in the schema file?

Author: For simplicity's sake, I was thinking of following the principles adopted by protobuf (forward-/backward compatibility rules) in terms of signals and resources, with the slight difference that the elements defined in the semantic convention registries (mainly attribute and metric definitions) follow the rules of transformation of the telemetry schema v1.1. Therefore, by default, a client SDK will be generated for the most recent version mentioned in the schema. However, it will be entirely possible for a backend component to consume data corresponding to an earlier version of the attributes and metrics, although in terms of signals and resources, apart from these elements, it will be necessary to be able to support the rules of classical forward-backward compatibility. For the moment, this seems to me a good balance between user experience and adaptability. This topic is clearly a debate to be had and experimented with.
It would be interesting to have @yurishkuro's feedback on this point, as a similar approach has been implemented at Meta. I am due to discuss this with him soon.

terms of telemetry.
* **Easy to exchange**: This artifact must be easily accessible from a web
server via a URL. This artifact must be small and avoid the repetition of
definitions.
* **Easy to parse**: A widespread and well-defined format should be preferred.
JSON is an example of such a format.
* **Easy to interpret**: The internal structure of this artifact must be
straightforward to avoid any misinterpretation.
* **Platform- and Language-agnostic**: This artifact must be independent of any
platform architectures and programming languages.

Conceptually, the main components of a Resolved Telemetry Schema are described
in the following diagram.

![Resolved Telemetry Schema](./img/0240-resolved-telemetry-schema.png)

The internal catalog is used to define all the attributes and metrics in this
artifact. This design allows for the reuse of the same attributes or metrics
multiple times in different signals. It is expected to be a very common pattern
to reuse the same subset of attributes or metrics across several signals.

A dedicated OTEP will precisely define the structure and format of the Resolved
Telemetry Schema.

> ## Roadmap

Author: In fact, I started with the proof of concept but the repository is not yet public (project code name OTel Weaver). I am working on it.
 
I have already done a demo to some members of the community (mainly the semantic conventions SIG, @tigrannajaryan , @jsuereth , @jmacd , @pyohannes , @lmolkova) but I can organize another broader demo if you want. Maybe within a SIG? I'm just not sure which one would be the most appropriate.

> ## Roadmap

Reviewer 2: perhaps a demo to the full TC?

> ## Roadmap

Author: Good idea. I'll see how to organize this with the full technical committee.

> ## Roadmap

Reviewer 3: @lquerel Dec 6, 8am PT is a TC meeting if it works for you.

> ## Roadmap

Author: @tigrannajaryan Yes that works. Thanks

>
> Note 3: For all the elements that make up the Application Telemetry Schema, a
> general mechanism of annotation or tagging will be integrated in order to
> attach additional traits, characteristics, or constraints, allowing vendors
> and companies to extend the definition of concepts defined by OpenTelemetry.
>
> Note 4: Annotations and Tags can also be employed to modify schemas for
> diverse audiences. For example, the public version of a schema can exclude all
> signals or other metadata labeled as private. Similarly, elements can be
> designated as exclusively available for beta testers. These annotations can
> also identify attributes as PII (Personally Identifiable Information), and
> privacy policy enforcement can be implemented at various levels (e.g., in the
> generated client SDK or in a proxy).

Several OTEPs will be dedicated to the precise definition of the structure and
the format of Telemetry Schema v1.2. The rules for resolving overrides
(inheritance), external references, and conflicts will also be described in
these OTEPs. See the Roadmap section for a comprehensive list of these OTEPs.

> ## Resolved Telemetry Schema

Reviewer 2: It is unclear to me what the scope of the resolved schema is. Is it a combination of all possible schemas used in an enterprise? Or a strict subset used in a specific application? If the latter, how does that work with oss instrumentation libraries that could refer to arbitrary schemas and are not uniformly discoverable?

Reviewer 2: The difference is we're not trying to obtain an exhaustive schema for a given application. An application produces one or more datasets, eg an RPC instrumentation can produce RPC metrics dataset. We want each dataset to have a well defined schema, which is an easier problem. Then the question of all datasets for an app is a discovery problem, which can be solved via search, especially if the app/workload identity is semantically annotated in the dataset schemas. Eg find all metrics/logs/etc with attribute service=myservice.

Author: Added a new section discussing in more details these problematics.

represented as a list of key/value pairs or as hashmaps, resulting in memory
overhead. A simple struct or a set of well-typed function arguments will be more
efficient and less error-prone for representing this list of attributes.

Databases and RPC systems (e.g., Protobuf & gRPC) have already addressed some of
these issues with a schema-first approach. There is nothing to prevent adopting
a similar approach in the context of telemetry. This document discusses how to
apply a schema-first approach in the OpenTelemetry project and its implications
for the existing Telemetry Schema and Semantic Conventions.

The following diagram provides an overview of the relationships between the
various components, processes, and artifacts of what could be a typical
schema-driven end-to-end telemetry system in the future.

![Application Telemetry Schema Overview](./img/0240-app-telemetry-schema-overview.png)

The concept of the Application Telemetry Schema, as described in the previous
diagram, will play a central role and unlock numerous use cases (this is not an
exhaustive list):

* Telemetry Client SDKs can be automatically generated from an application

> telemetry schema, enhancing both user experience and performance.

Author: I agree that this diagram is not clear enough. The approach is as follows: the developer of an application or a service defines an application telemetry schema describing the telemetry signals emitted by that component (also true for third party libraries). From the application telemetry schema, a "compiler" (a la protoc) produces a resolved telemetry schema that will be used to generate an Client SDK/API optimized for that specific schema. This same resolved telemetry schema can be used for other functions in the CI/CD pipeline (e.g., privacy policy enforcement, compatibility checker) or in a telemetry backend (e.g., dashboard generation, table schema generation, schema-driven data processing, ...). I will update the diagram accordingly.

> telemetry schema, enhancing both user experience and performance.

Author: Udpated first diagram. Let me know if it's not clear enough.

schemas and semantic convention registries. It involves several vendor and
enterprise artifacts, in addition to the standardized OpenTelemetry artifacts.
The schema resolution process will produce a self-contained Resolved Telemetry
Schema that can be easily consumed by various tools and applications, such as a
Client SDK generator, compatibility checker, compliance checker, data catalog
feeder, and more.

![Example of Application Telemetry Schema Hierarchy](./img/0240-app-telemetry-schema-hierarchy-example.png)

For each important component, the following diagram defines the responsibilities
and key design properties that have been considered in the subsequent sections
of this document.

![Application Telemetry Schema - Levels of responsibilities & Key design properties](./img/0240-app-telemetry-schema-design-properties.png)

This design enables the definition of semantic conventions in a distributed
manner. OpenTelemetry, vendors, and enterprises can define their own semantic
conventions in different registries simplifying the existing process.

The next two sections will describe the concepts of Telemetry Schema v1.2 and
the Resolved Telemetry Schema.

> ## Telemetry Schema v1.2 (aka Application Telemetry Schema)

Reviewer 4: Currently, the telemetry schemas [v1.1.0 file format](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/file_format_v1.1.0.md) only defines the changes in the schema from version to version. Presumably, you aim to expand this to be able to declare the schema itself, and not just the changes. If so, big 👍 from me.

Author: Yes this is exactly the intent.

  than 2 levels. This is useful for supporting complex environments and will
  allow the distribution of semantic convention registries across vendors and
  enterprises. This OTEP depends on the “Schema Child/Parent relationship
  resolution” OTEP.
* Resolved Schema Format v1.0.
* Schema-driven Client SDK Generator.
* Schema Child/Parent relationship resolution.

### Tools

OpenTelemetry tools

* Semantic Convention Registry Merger (Johannes Tax's proposal): This tool will
merge all the semantic convention files of a registry into a single and
referencable file.
* Search Engine: This tool will provide search capability within semantic
convention registries and telemetry schemas.
  * V1 focusing on the semantic convention registry.
  * V2 focusing on telemetry schema v1.2.
* Telemetry Schema Resolver: This tool will resolve the telemetry schema and its
dependencies. The main outcome of this tool is a Resolved Telemetry Schema.
* Schema-driven Client SDK Generator: This tool will generate a client SDK or a
client API from a telemetry schema v1.2. It will be designed to be extensible,
allowing each language SIG and the general community to create templates for

> generating client SDKs and APIs in specific programming languages.

Reviewer 4: This is conceptually similar to the java [instrumentation-api](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation-api). The docs on instrumentation-api are sparse, but the general idea is to provide a higher order API built on top of the metric, trace, log API (only metric and trace in practice) which makes it easy to build instrumentation which conforms to the semantic conventions. As an instrumentation author, you implement a hand full of interfaces to extract attributes relevant to the semantic convention being implemented, and the instrumentation API takes care of recording spans / metrics correctly.

Author: I may not be familiar enough with the Java Instrumentation API, but my understanding so far was that the Instrumentation API contains all the attributes and metrics defined in the official semantic convention registry. So, in a way, it's not code generated for a specific application context (or library) but rather code generated for a registry or a catalog of conventions. If my interpretation is correct, then I think that this OTEP follows a somewhat different approach, although complementary.

* Telemetry Schema Compatibility Checker: This tool will verify that two
successive versions of the same Telemetry Schema adhere to the specified

> compatibility rules.

Reviewer 4: Would be good to have some sort of component that could validate that incoming telemetry conforms to the telemetry schema it uses to describe itself. For a while, I've imagined a collector processor which inspects data and says whether or not the data conforms to the semantic conventions of its schemaUrl. This effort could be a good chance to build such a tool.

Generally, I do think its important that we have at least some sort of tool that consumes telemetry schemas. Building this tool confirms that the concepts in telemetry schema are suitable to build on top of, and acts as a reference implementation for how other consumers of telemetry schemas might work.

Author: Indeed, one of the objectives of this project is to be able to validate incoming telemetry among other things (e.g., privacy policy enforcement, dashboard generation, database schema generation, schema-driven data transformation, ...).

I have just made public this [repo](https://github.com/f5/otel-weaver) containing a tool (PoC) that explores and implements some aspects of this OTEP.

On the application and service owners' side, they must either make a local copy
of the resolved telemetry schemas corresponding to their dependencies (similar
conceptually to the model followed by protobuf and gRPC) or link to the resolved
telemetry schema for a specific version. In both cases, the synchronization
between the versions of the library and the resolved telemetry schema is the
responsibility of the application owner.

![Centralized Approach](./img/0240-centralized-approach.png)

This approach is considered centralized as the application owner has the
responsibility of maintaining, in a centralized manner locally, all the
telemetry schemas of the dependencies.

A specific OTEP dedicated to the resolved telemetry schema will be responsible
for defining the format and structure of the resolved telemetry schema, as well
as the best approach to collect the resolved telemetry schemas of the different
dependencies.

## Roadmap

To facilitate the review process and progressively deliver value to the
OpenTelemetry community, a series of OTEPs and tools are proposed.

### OTEPs

* Telemetry Schema versions (incremental steps):

>   * V1.1.1: Add a section to import one or several Semantic Convention

Reviewer 3: I would like to discuss in a separate OTEP the exact structure of the schema files that "import" vs schema files that are called "resolved schemas" in this OTEP. I think it is important that we always publish the resolved schemas that don't require further link resolution to be useful for processing purposes. We don't have to discuss the details here, just wanted to call out that the ability "import" does not necessarily need to be a feature of the published schema files.

Author: I see. This is also related to your subsequent comment about considering the current telemetry schema version 1.1.0 as actually being a precursor to what I refer to here as the resolved telemetry schema. I need to go through the entire document again to make the necessary adjustments.

Reviewer 5: > I think it is important that we always publish the resolved schemas

+1

Author: I have added an 'Open Questions' section at the end of this OTEP. The purpose is to clearly expose the important questions that will need to be resolved in the following stages/OTEPs. The current OTEP focuses on the overall vision and the major concepts of this initiative. One of the questions is about having one or two types of schema and also includes a format aspect for the schema(s). The point discussed in your comment will be covered in the context of responding to this question in the next OTEP (see roadmap), which I hope to start soon.

  Registries. Once implemented, the official Semantic Convention Registry will
  be referenced from the official OpenTelemetry Root Telemetry Schema.
  * V1.2: Add references to parent schemas and define a schema section to
  describe application/library signals. In this version, only metrics, logs, and
  spans signals are described. This is the first version of the telemetry schema

>   that can be used by the Client SDK generator. In this version only one schema

Reviewer 3: >This is the first version of the telemetry schema
>  that can be used by the Client SDK generator.

Why can't V1.1.1 be used by the generator? I think it will contain enough information to generate code for Otel-wide conventions.

Author: See https://github.com/open-telemetry/oteps/pull/243/commits/b1954142ff280067f16185d3f7c89e7985902bc3

  can be inherited. More complex hierarchies will be supported in version 1.5.
  * V1.3: Add support for events in the schema section. This OTEP is contingent
  upon the approval of events in OpenTelemetry.
  * V1.4: Add support for metric_groups (multivariate metrics) in the schema
  section. This will only be relevant if a support for multivariate metrics is
  planned in the Client SDK Generator.
  * V1.5: Add support for multi-inheritance and for schema hierarchies of more
  than 2 levels. This is useful for supporting complex environments and will
  allow the distribution of semantic convention registries across vendors and
  enterprises. This OTEP depends on the “Schema Child/Parent relationship
  resolution” OTEP.

> * Resolved Schema Format v1.0. This OTEP will define both the structure and

Reviewer 3: I would like to consider flipping this the other way round. 

Instead of introducing a new concept of "resolved schema" I want to say that what we have today _is_ the resolved schema. That's what we publish today and what we will continue publishing in the future (with more features in newer versions). The "unresolved" schema is the new concept that we will introduce which possibly may even exist only internally at Otel and not intended for public consumption (unless we see good reasons for such consumption). The unresolved schema + semantic conventions will be the input to an internal tool that produces the Schema Files as we know them (and which are always resolved).

Author: Yes, I understand the rationale and it doesn't seem to be a significant modification. It will require me to make some minor adjustments here and there in the OTEP. I will take care of that as soon as possible.

Reviewer: I fully agree with @tigrannajaryan, this also takes care of https://github.com/open-telemetry/oteps/pull/243/files#r1413837539.

Publishing resolved schemas would allow us to go without requiring a public specification of the attribute registry, which would be required if users were required to resolve schemas. The schema being the single point where users interface with semantic conventions makes life easier for users, as well as for spec authors.

Schema is created from a Component Telemetry Schema. This **Resolved Telemetry
Schema is self-contained** and has no external references. Optionally, a Component
Telemetry Schema may extend an existing Telemetry Schema, whether component or
resolved (this aspect is still under discussion). Typically, the official
OpenTelemetry Telemetry Schema is inherited by a Component Telemetry Schema to
include the standard OpenTelemetry Semantic Convention registry. In complex
cases, large enterprises might create their own intermediary telemetry schemas
for custom definitions.

Each signal definition defined in the Component Telemetry Schema, where possible,
reuses the existing syntax and semantics defined by the semantic conventions.
Each signal definition is also identified by a unique name (or ID), making
schemas addressable, easy to traverse, validate, and diff.

The key design principles to be followed in the definition of the Resolved
Telemetry Schema are:

* **Self-contained**: No external references are allowed. This artifact contains
  everything required to determine what an application or a library produces in
  terms of telemetry.

> * **Easy to exchange**: This artifact must be easily accessible from a web
>   server via a URL. This artifact must be small and avoid the repetition of

Reviewer 2: ```suggestion
* **Easy to exchange**: This artifact must be easily accessible to actors
  via a URI. This artifact must be small and avoid the repetition of
```

Reviewer 2: "web server" is too specific, doesn't need to be used (e.g. maybe schemas can be loaded from a jar)

![Use cases](./img/0240-otel-weaver-use-cases.svg)

Note: The relationship between Telemetry Schema v1.1
([OTEP 0152](https://github.com/open-telemetry/oteps/blob/main/text/0152-telemetry-schemas.md))
and the Component and Resolved Telemetry Schema concepts is still being
discussed. This will be clarified in future OTEPs (refer to the last section).

The following diagram illustrates a possible instance of a complex hierarchy of
schemas and semantic convention registries. It involves several vendor and
enterprise artifacts, in addition to the standardized OpenTelemetry artifacts.
The schema resolution process will produce a self-contained Resolved Telemetry
Schema that can be easily consumed by various tools and applications, such as a
Client SDK generator, compatibility checker, compliance checker, data catalog
feeder, and more.

![Example of Telemetry Schema Hierarchy](./img/0240-otel-weaver-hierarchy.svg)

This hierarchy of telemetry schemas helps large organizations in
collaborating on the Component Telemetry Schema. It enables different
aspects of a Component Telemetry Schema to be managed by various teams.

For all the elements that make up the Component Telemetry Schema, a

> general mechanism of annotation or tagging will be integrated in order to

Reviewer 2: > annotation or tagging

I understand this is vague on purpose, but one thing I'd want to avoid is having two independent DSLs, one for schema definition and another for annotations. If you noticed in Meta's schema-first telemetry paper, the annotations and the schema both used the exact same Thrift IDL mechanism (which, in turn, means annotations can be arbitrarily nested).

Author: Yes we are on the same page. I will clarify this part of the OTEP.

Author: Added the following sentence: This annotation mechanism will be included as part of the Component Telemetry Schema definition.

questions is not critical at this stage and will likely become clearer as we
define and implement future OTEPs.

### Should we use a single Telemetry Schema or the combination of Component and Resolved Telemetry Schema?

The debate between adopting a single Telemetry Schema or separate Component and
Resolved Telemetry Schemas remains unresolved.

Advocates of a single-schema approach see it as a simplification in terms of
schema definition, implementation, and even cognitive overhead. They suggest
using the same schema but disallowing any external references for schemas
intended for publication and reuse by third parties. The schema resolution
process would then remove these references.

Proponents of a two-schema approach believe that each format is intended for
different users and use cases (app/lib developers vs telemetry tool developers),
and therefore, having two distinct structures and formats would make it easier
to optimize each for its specific use case (in multiple dimensions). The first
group, app and lib developers, is much larger than the second and would not need
to understand the details of the second format.

> ### What should be the schema(s) be named?

Reviewer 2: This is misleading, you can be asking about conceptual names (component vs. resolved schemas), or how they are implemented (same DSL or different)

Author: I have updated this sentence as follows: 
> Furthermore, the group of developers using the Component Telemetry Schema will most likely be much larger than the group of developers who need to understand the details of the Resolved Telemetry Schema. 

I hope it's less misleading.

ecosystem. It proposes extending the initiatives of Telemetry Schema and
Semantic Conventions to include logical concepts of Component Telemetry Schema
and Resolved Telemetry Schema. A series of OTEPs and Tools will be proposed in
this overarching document to detail each aspect of this vision._

## Current State Overview

Traditionally, the instrumentation of applications is deeply integrated into the
source code of the applications and their components. The current stack of
OpenTelemetry follows this approach and offers a unified mechanism that allows
this ecosystem to report telemetry data jointly via a generic client interface
and a common protocol through an SDK. Moreover, OpenTelemetry's semantic
conventions establish a vendor-agnostic and standardized vocabulary to ensure
more consistent instrumentation over time. This standardization facilitates the
analysis and utilization of these metadata data across the entire telemetry
pipeline.

But this approach is not without challenges:

* **Discoverability and Interoperability**: It is difficult to discover a priori
  and automatically what an application as a whole specifically generates in terms

>   of telemetry and associated metadata.This makes it difficult to integrate with

Reviewer 6: ```suggestion
  of telemetry and associated metadata. This makes it difficult to integrate with
```

  library's development cycle. These metadata definitions should be distinct from
  the signal definitions and could be specified by teams other than development
  teams.
* **Performance overheads**: A significant downside of generic telemetry
  instrumentation is the various overheads it generally introduces due to inherent
  layers of abstraction. For example, the collection of attributes is typically
  represented as a list of key/value pairs or as hashmaps, resulting in memory
  overhead. A simple struct or a set of well-typed function arguments will be more
  efficient and less error-prone for representing this list of attributes. In the
  same way, it is possible to use a dictionary encoding for values whose domain is
  specified in the form of an enumeration in the schema definition.

Databases and RPC systems (e.g., Protobuf & gRPC) have already addressed some of
these issues with a schema-first approach. There is nothing to prevent adopting
a similar approach in the context of telemetry. **This document discusses how to
apply a schema-first approach in the OpenTelemetry project and its implications
for the existing Telemetry Schema and Semantic Conventions.**

> The following diagram provides a conceptual overview of the relationships

Reviewer 2: this still goes under Current State header, a bit confusing, maybe add "Possible/Desired Future State"

> diverse audiences. For example, the public version of a schema can exclude all
> signals or other metadata labeled as private. Similarly, elements can be
> designated as exclusively available for beta testers. These annotations can
> also identify attributes as PII (Personally Identifiable Information), and
> privacy policy enforcement can be implemented at various levels (e.g., in the
> generated client SDK or in a proxy).
>
> Note 5: This
> recent [paper](https://arxiv.org/pdf/2311.07509.pdf#:~:text=The%20results%20of%20the%20benchmark%20provide%20evidence%20that%20supports%20our,LLM%20without%20a%20Knowledge%20Graph)
> from [data.world](https://data.world/home/), along with
> the [MetricFlow framework](https://docs.getdbt.com/docs/build/about-metricflow)
> which underpins the [dbt Semantic Layer](https://www.getdbt.com/product/semantic-layer),
> highlights the significance of adopting a schema-first approach in data
> modeling, especially for Generative AI-based question answering systems. Tools
> like Observability Query Assistants (
> e.g. [Elastic AI Assistant](https://www.elastic.co/fr/blog/introducing-elastic-ai-assistant)

> >
> and [Honeycomb Query Assistant](https://www.honeycomb.io/blog/introducing-query-assistant?utm_source=newswire&utm_medium=link&utm_campaign=query_assistant))

Reviewer 2: ```suggestion
> and [Honeycomb Query Assistant](https://www.honeycomb.io/blog/introducing-query-assistant?utm_source=newswire&utm_medium=link&utm_campaign=query_assistant))
```

  and cons of a single schema versus a dual-schema approach. It aims to identify
  the optimal solution and define the structures and formats for the two
  concepts introduced in this OTEP: the Component and Resolved Telemetry Schema.
  We plan several sub-OTEPs to progressively introduce and phase the concepts
  discussed in this document:
  * **Phase 1**: Introduce a registry section with a list of fully
  resolved attributes, metrics, and other telemetry signals from the existing
  semantic convention registry. Upon implementation, the OpenTelemetry project's
  telemetry schema will include a registry of all standardized attributes,
  metrics, and signals as defined by the semantic conventions.
  * **Phase 2**: Add sections for resource, instrumentation library,
  and schema sections to represent telemetry signals emitted by an application
  (and its dependencies) or a library. Initially, only metrics, logs, and spans
  definitions will be supported.
  * **Phase 3**: Enable application or library authors to
  override definitions inherited from a parent resolved schema.
  * **Phase 4**: Implement support for events in the schema

>   section, pending the approval of events in OpenTelemetry.

Reviewer 2: Is this included because you consider it a new signal type? It feels somewhat irrelevant to this project, a specific sub-domain that will simply follow the same rules as the other sub-domains.

Author: I have included events in a separate phase for the following reasons:
- It's a new type of signal that I believe has not yet been approved.
- If it is approved, and although it is protocol-compatible with log-type signals, it will be necessary to reify this concept at the telemetry schema level to make the most of it.

> from [data.world](https://data.world/home/), along with
> the [MetricFlow framework](https://docs.getdbt.com/docs/build/about-metricflow)
> which underpins the [dbt Semantic Layer](https://www.getdbt.com/product/semantic-layer),
> highlights the significance of adopting a schema-first approach in data
> modeling, especially for Generative AI-based question answering systems. Tools
> like Observability Query Assistants (
> e.g. [Elastic AI Assistant](https://www.elastic.co/fr/blog/introducing-elastic-ai-assistant)
> and [Honeycomb Query Assistant](https://www.honeycomb.io/blog/introducing-query-assistant?utm_source=newswire&utm_medium=link&utm_campaign=query_assistant))
> are likely to become increasingly prevalent and efficient in the near future,
> thanks to the adoption of a schema-first approach.

Several OTEPs will be dedicated to the precise definition of the structure and
the format of this/these schema(s). The rules for resolving overrides
(inheritance), external references, and conflicts will also be described in
these OTEPs. See the Roadmap section for a comprehensive list of these OTEPs.

## Resolved Telemetry Schema

> A Resolved Telemetry Schema is the outcome of the schema resolution process.

Author: Indeed. I will make this change in the OTEP.

> A Resolved Telemetry Schema is the outcome of the schema resolution process.

Author: Done. See https://github.com/open-telemetry/oteps/pull/243/commits/87840e2533ff1f89955909e9e40438406982a45d

This process involves taking the entire hierarchy of Telemetry Schemas and
Semantic Convention Registries and applying a set of rules to resolve overrides
and eliminate external references. The key design principles to be followed in
the definition of the Resolved Telemetry Schema are:

* **Self-contained**: No external references are allowed. This artifact contains
  everything required to determine what an application or a library produces in
  terms of telemetry.
* **Easy to exchange**: This artifact must be easily accessible from a web
  server via a URL. This artifact must be small and avoid the repetition of
  definitions.
* **Easy to parse**: A widespread and well-defined format should be preferred.
  JSON is an example of such a format.
* **Easy to interpret**: The internal structure of this artifact must be
  straightforward to avoid any misinterpretation.
* **Platform- and Language-agnostic**: This artifact must be independent of any
  platform architectures and programming languages.

The following diagram describes two main use cases for the Resolved Telemetry
Schema. The key points to remember are: 1) both use cases result in a Resolved

> Telemetry Schema, and 2) Resolved Telemetry Schemas serve as the mechanism for

Author: Exact that's the idea.

> Telemetry Schema, and 2) Resolved Telemetry Schemas serve as the mechanism for

Author: Updated the OTEP to clarify this point. See https://github.com/open-telemetry/oteps/pull/243/commits/9369b8a23049bb752078291daa30fcd7aad4eca8

distributing Telemetry Schemas throughout the entire ecosystem.

> ![Use cases](./img/0240-otel-weaver-use-cases.svg)

Reviewer: Is it required for end-users to use a Component Telemetry Schema? Or could they just use the Resolved Telemetry Schema provided by OTel, assuming that no customizations are needed?

Author: Yes it is conceivable to use a resolved telemetry schema, representing a compact and resolved version of a semantic convention registry, in a manner similar to that of current SDK/API clients, where nothing specific is specified/known.
However, the preferred mode (or main purpose) of use for the component telemetry schema in this proposal is to define the telemetry objects produced by an application or a library.

The main components of a Resolved Telemetry Schema are illustrated in the
diagram below. The 'OTel Weaver' tool is used to create these schemas. It can
also extend an existing schema or import a Semantic Convention Registry.
Resolved Telemetry Schema serves as a key mechanism for interoperability,
feeding various external tools, including SDK generators, documentation
generators, policy enforcers, and more.

> ![Resolved Telemetry Schema](./img/0240-otel-weaver-resolved-schema.svg)

Reviewer: Nit: according to the diagram, a Resolved Telemetry Schema refers to _exactly one_ instrumentation library. Will this also true for the Resolved Telemetry Schema created directly from OpenTelemetry semantic conventions and provided by OpenTelemetry?

Author: Not exactly. The dependencies section will contains the instrumentation libraries of each dependency (see the appendix B I recently added to this OTEP). Regarding a resolved telemetry schema representing a semantic convention registry, the instrumentation library is not defined.

or metrics across several signals and libraries.

A dedicated OTEP will precisely define the structure and format of the Resolved
Telemetry Schema.

### Development Strategies

Two development strategies coexist and must be supported. The first strategy, a
monorepo type (single source of truth), has complete control over the
applications, their dependencies, and the associated telemetry schemas. The
second strategy is more heterogeneous, composed of multiple repositories, where
the build process of external dependencies is out of the control of the entity
owning the applications or services.

In the first model, each build process can independently apply telemetry schema
policies in parallel, knowing that the policies are shared and the entire
environment is controlled by the same entity.

In the second model, the application or service environment does not control the
telemetry schema policies of external dependencies. There is a need for a method
to retrieve telemetry schemas from these dependencies. The mechanism for
distributing these schemas is still being discussed (refer to the Open Questions
section). Ultimately, this will enable the CI/CD pipeline of the application or

> service to apply its own local policies to these schemas from the dependencies.

Reviewer: It's not clear to me how this should work. 

I guess this might work if those "local policies" only have an effect on the SDK, but I assume they can't change any API that is used by the dependencies?

Anyway, this is covered by an open question, and doesn't need to be answered in this OTEP.

Author: Indeed, this will be covered in a future OTEP.

# This file describes the structure and format of a component telemetry schema.
# A component telemetry schema can be used for three types of purposes:
# - defining the telemetry schema for an application or a service
# - defining the telemetry schema for a library
# - defining a semantic convention registry
file_format: 1.2.0
schema_url: <url-of-the-current-component-schema>
# This optional field allows to specify the parent schema of the current schema.
# The parent schema is a resolved schema.
parent_schema_url: <url-of-a-parent-resolved-schema>

# This optional section allows for importing a semantic convention registry
# from a git repository containing a set of semantic convention files. It is
# also possible to import file by file.
semantic_conventions:
  - git_url: <git-url-of-the-semantic-conventions-repository>
    path: <path-to-the-semantic-conventions-directory-inside-the-git-repo>
  - url: <url-of-the-semantic-conventions-file>

# The resource field is defined when the component schema is that of an
# application (as opposed to that of a library). The resource field contains a
# list of local references to attributes defined in the shared catalog within
# this file.

> resource:

Reviewer 2: This is interesting. When I think of a "component schema" I think of a compile unit, i.e. the schema of the telemetry is "fixed" after the code is compiled. But the resource attributes are not scoped to source code, they can be quite often added dynamically depending on where that compiled unit is deployed. Strictly speaking, you can even say that about other attributes, e.g. a processor in the collector could add certain attributes to telemetry (e.g. "processing timestamp"). So the resulting schema seen at rest may not match the component schema. 

How are the dynamic enrichments should be accounted for in the schema management lifecycle?

Author: For this specific case, I was thinking of offering the possibility to define the value of an attribute dynamically by referencing an environment variable (or any other value injection mechanism at the time of deployment). Thus, the attribute is defined in the schema with a clearly identified placeholder. The value of this placeholder is resolved at the time of deployment or by the SDK at runtime in a particular environment. We can also support cases where the identifiers of the attributes themself are not known before deployment, but in this case, these attributes will not be controlled/known at the time of the build.

Reviewer 2: but I am not talking about values, rather the actual attributes. If your app is deployed on AWS, it may get AWS-specific resource attributes, which it wouldn't get on GCP.

Author: Yes, I had a doubt, which is why the second part of my response includes the case you are talking about, that is, the case where certain attributes (identifier and value) are only known at the time of deployment.

Reviewer 2: seems like it's often going to be the case, so if the ingestion pipeline does any kind of schema validation it would need to know these deployment-enriched materialized schemas
